airflow/www/static/js/dag/details/task/TaskDuration.tsx

```diff
@@ -179,7 +170,18 @@ const TaskDuration = () => {
     // @ts-ignore
     dataset: {
       dimensions: ["runId", "queuedDurationUnit", "runDurationUnit"],
-      source: durations,
+      source: durations.map((duration) => {
+        if (duration) {
+          const durationInSeconds = duration as TaskInstanceDuration;
```

Comment:
I couldn't find the right way to convince typescript compiler here since duration is a union type of {} | TaskInstanceDuration. Neither checking for duration to be true in if clause nor using (duration as TaskInstanceDuration) / unitDivisor without a temporary variable.

Reply:
Actually we can drop the `<RunDuration | {}>[]` and just have `RunDuration[]`


And for taskDurations we can change from `map` to `forEach` like so:

``` 
const durations: TaskInstanceDuration[] = [];

  dagRuns.forEach((dagRun) => {
    const { runId } = dagRun;
    const instance = task.instances.find((ti) => ti && ti.runId === runId);
    if (instance) {
 ```
 
 and then
 
 ```
      durations.push({
        ...instance,
        [orderingLabel]: dagRun ? dagRun[orderingLabel] : instance.startDate,
        runDuration,
        queuedDuration,
        runDurationUnit,
        queuedDurationUnit,
      });
```

Reply:
In the initial PR it used to be that way using .push and foreach but I remember vaguely that eslint had something about foreach and change was made to map using TaskInstanceDuration[] and then null was added that later changed to {}. Some of the relevant diffs as below. I guess this was made for a reason. I would keep the cast as it is a simpler change to typecast like assert and use the object. Thanks.

Relevant commits : 

https://github.com/apache/airflow/commit/8156faa5f710b0e21ab741073d2a889e019240e6
https://github.com/apache/airflow/commit/23435af92dcaa4f0668978dbf103706d7c02acb0

```diff
-  dagRuns.forEach((dagRun) => {
-    const instance = hasTaskInstances.instances.find(
-      (_instance) =>
-        _instance.taskId === taskId && _instance.runId === dagRun.runId
-    );
-
-    if (!instance) return null;
+  const durations: TaskInstanceDuration[] = task.instances.map((instance) => {
     // @ts-ignore
     const runDuration = moment.duration(
       instance?.startDate && instance?.endDate
@@ -108,77 +73,77 @@ const TaskDuration = () => {
       maxDuration = runDuration.asSeconds();
     }
 
-    durations.push(
-      new TaskInstanceDuration(
-        instance,
-        runDuration,
-        queuedDuration,
-        dagRun.dataIntervalStart,
-        dagRun.dataIntervalEnd,
-        dagRun.executionDate
-      )
-    );
-
-    return null;
-  });

```


```
-  const durations: TaskInstanceDuration[] = task.instances.map((instance) => {
+  const durations: (TaskInstanceDuration | null)[] = dagRuns.map((dagRun) => {
+    const { runId } = dagRun;
+    const instance = task.instances.find((ti) => ti && ti.runId === runId);
+    if (!instance) return null;
```

```
diff --git a/airflow/www/static/js/dag/details/task/TaskDuration.tsx b/airflow/www/static/js/dag/details/task/TaskDuration.tsx
index 1f99f39e4d..cb6ad1f69d 100644
--- a/airflow/www/static/js/dag/details/task/TaskDuration.tsx
+++ b/airflow/www/static/js/dag/details/task/TaskDuration.tsx
@@ -56,10 +56,10 @@ const TaskDuration = () => {
   if (!task) return null;
   const orderingLabel = ordering[0] || ordering[1] || "startDate";
 
-  const durations: (TaskInstanceDuration | null)[] = dagRuns.map((dagRun) => {
+  const durations: (TaskInstanceDuration | {})[] = dagRuns.map((dagRun) => {
     const { runId } = dagRun;
     const instance = task.instances.find((ti) => ti && ti.runId === runId);
-    if (!instance) return null;
+    if (!instance) return {};
```

https://github.com/github/eslint-plugin-github/blob/main/docs/rules/array-foreach.md

Reply:
Yes, my bad. I remember now why we changed that. We needed empty columns in case there was no task instance for a particular dag run. Let's leave you work as-is then